Qubes 4.0 menu support

3hhh opened the pull request:
This PR

- makes the application menu work again in Qubes 4.0. The python code from 3.2 didn't work anymore as the Qubes API changed.

- adds an optional awesome-specific application menu which can be created with xdg menu files at ~/.config/awesome/xdg-menu/

- does some cleanup

Further details can be found in the respective commit messages.

I attempted to keep the changes of PR #4  alive, but the function parameters had to be changed unfortunately due to the Qubes API changes.

3hhh commented:
Not sure why code-signing failed; git shows everything verified.

marmarek commented:
> Not sure why code-signing failed; git shows everything verified.

Not sure either. Do you have your key available on some keyserver?

marmarek commented on qubes.lua.in:
```diff
@@ -170,60 +161,61 @@ function qubes.get_colour_focus(c)
-    return {'Qubes Manager', 'qubes-manager',
-        '/usr/share/icons/hicolor/16x16/apps/qubes-logo-icon.png'}
+function is_accessible(path)
+    return os.rename(path, path) and true or false
```
Trying to rename the file just to check its existence looks very wrong.

3hhh replied:
I agree. However it seems that standard lua doesn't support a straightforward function to get the job done without additional dependencies.
A few alternatives: https://stackoverflow.com/questions/1340230/check-if-directory-exists-in-lua

Attempting to read or call python/bash/whatever might be another alternative.
I picked rename as it should be fast & even atomic most of the time; it also provides a free permission check.
Just let me know which one you prefer.

marmarek replied:
@woju any idea?

marmarek commented on qubes.lua.in:
```diff
@@ -170,60 +161,61 @@ function qubes.get_colour_focus(c)
+--all parameters except for vmname are optional
+function qubes.make_vm_menu(vmname, menuitemfunc, vmapps, vmicon)
     local menu = {}
+    local vmapps = vmapps or HOME .. '/.local/share/qubes-appmenus/' .. vmname .. '/apps/'
```
Why don't use files installed in standard location: `~/.local/share/applications`? Or more generally - why this code is needed at all - does Awesome not support standard menu from `.desktop` and `.directory` and `.menu` files (i.e. XDG standard)?

Using standard location will work on any Qubes version.

3hhh replied:
Honestly I just continued to do what was previously done in this repo. I think the approach was implemented as awesome supports `.desktop` files, but no `.menu` or `.directory` files.
@woju Can you confirm my assumption?

In 3.2 `/var/lib/qubes/[vm]/apps` was parsed by the `qubes.lua` code. The VM names were used as the first level of menu items. In 4.0 that structure was apparently moved to `~/.local/share/qubes-appmenus/`, so I used that one.

So yes, we can just use `~/.local/share/applications`, but that would then create menu entries which are all on the first level and create a rather messy menu as all imposed hierarchy from the `.menu` files is lost.

It doesn't look like there are plans by awesome to support `.menu` files any time soon - at least I didn't find any related issue on their bugtracker. By default they seem to support categories for some menu structure [1]. That's also what other menu implementations seem to use ( e.g. [2] [3]), but isn't really helpful for a VM-based hierarchy.

Some proper `.menu` file parsing would of course be the best solution, but require quite a lot of work I guess.

If you don't want to stick to the old approach, I'd recommend to just reject this PR.

[1] https://awesomewm.org/apidoc/libraries/menubar.menu_gen.html
[2] https://wiki.archlinux.org/index.php/Xdg-menu
[3] https://github.com/lcpz/awesome-freedesktop

marmarek replied:
I'm trying to avoid the need to port this code to every new Qubes version. Using `.menu` files should accomplish it, even if just minimal support is here. I don't know lua (and its support for XML), but for shell something like this would work:

For each menu file:
1. Get menu name:
```
xmllint --xpath '/Menu/Menu/Name/text()' /path/to/file.menu
```
2. Get list of included desktop files:
```
echo '/Menu/Menu/Include/Filename/text()' | xmllint --shell /path/to/file.menu | grep 'desktop$'
```

Then, look for those files in `~/.local/share/applications/`. Rest of the code should be the same.

3hhh replied:
So essentially you suggest to parse the `.menu` files at `~/.config/menus/applications-merged` and use the directory mentioned in them as 1st level menu, then parse the `.desktop` files at `~/.local/share/applications` and use the entries there as 2nd level menu?

Maybe that's easily feasible with some lua regex, I'll check.

marmarek replied:
> So essentially you suggest to parse the .menu files at ~/.config/menus/applications-merged and use the directory mentioned in them as 1st level menu, then parse the .desktop files at ~/.local/share/applications and use the entries there as 2nd level menu?

Yes, exactly.

3hhh replied:
I changed that now, but didn't test too much yet.

marmarek commented on qubes.lua.in:
```diff
@@ -170,60 +161,61 @@ function qubes.get_colour_focus(c)
+ret = []
+for vm in app.domains: ret.append(vm.name)
+ret.sort(key=lambda i: i.lower())
+print('\n'.join(ret))
```
Better use `qvm-ls --raw-data -O name` to not duplicate that code.
Or `qvm-ls --raw-list`, as soon as I'll bring it back to Qubes 4.0. This way it will work for both 3.2 and 4.0.

3hhh replied:
I could also use a straightforward `ls ~/.local/share/qubes-appmenus/`. Or include some library to be able to read directory contents...
But I'll wait for your call wrt the overall future approach.

3hhh commented:
> Not sure either. Do you have your key available on some keyserver?

No, I dont think so. Probably that's why then.

3hhh commented:
@marmarek I now squashed a more thoroughly tested version to the last commit. From my point of view everything is now implemented as we discussed except for the `os.rename(path, path)` line, i.e. I'd call it ready for a 2nd round of review.

Unfortunately the change of approach basically resulted in quite some lines...

Anyway it is now also compatible to 3.2 except for the changes of PR #4 as mentioned before.

I also updated the PR description.

marmarek commented:
Thanks!
